### PR TITLE
Use tiny models for ONNX tests

### LINKED ISSUE #18819: ONNX test suite is slow - run in 5.5 hours
### Who can help?

@lewtun  @LysandreJik 

As shown in [this job](https://github.com/huggingface/transformers/runs/8074936385?check_suite_focus=true) run, the ONNX tests now run in 5.5 hours.

From https://github.com/huggingface/transformers/blob/73c6273d481f9052098a2a5a5f001fa75daaace9/tests/onnx/test_onnx_v2.py#L182
we see that the tests use real model checkpoints. As ONNX graph compile is known to be slow, and the models are quite big, it makes the tests very slow.

The whole scheduled CI test suite now run in 14.5 hours, and we have 2 test suites to run each day, so it requires 29 hours and can't finish in one day. This causes the test suite and their reports being delayed a lot.

We are wondering if it is possible to use tiny models from [hf-internal-testing](https://huggingface.co/hf-internal-testing) for ONNX tests.

### PR DESCRIPTION
**Pull request by lewtun**
# What does this PR do?

Uses tiny random models for the ONNX tests to speed up the test suite. Closes #18819 

## TODO

- [x] Add tiny model for `deepmind/language-perceiver`
- [x] Add tiny model for `deepmind/vision-perceiver-conv`
- [x] Add tiny model for `hustvl/yolos-tiny`
- [x] Add tiny model for  `nvidia/segformer-b0-finetuned-ade-512-512`
- [x] Add tiny model for `google/long-t5-local-base`
- [ ] Ensure slow tests pass

### Hub PRs that need merging to ensure slow test pass

- [x] https://huggingface.co/hf-internal-testing/tiny-random-beit/discussions/1
- [x] https://huggingface.co/hf-internal-testing/tiny-random-deit/discussions/1
- [x] https://huggingface.co/hf-internal-testing/tiny-random-deit/discussions/2
- [x] https://huggingface.co/hf-internal-testing/tiny-random-clip/discussions/1
- [ ] https://huggingface.co/hf-internal-testing/tiny-random-clip/discussions/2
- [x] https://huggingface.co/hf-internal-testing/tiny-random-convbert/discussions/1
- [x] https://huggingface.co/hf-internal-testing/tiny-random-xlm-roberta/discussions/1
- [x] https://huggingface.co/hf-internal-testing/tiny-random-xlm-roberta/discussions/2
- [x] https://huggingface.co/hf-internal-testing/tiny-random-ibert/discussions/1
- [x] https://huggingface.co/hf-internal-testing/tiny-random-ibert/discussions/2
- [x] https://huggingface.co/hf-internal-testing/tiny-random-blenderbot-small/discussions/1
- [x] https://huggingface.co/hf-internal-testing/tiny-random-blenderbot-small/discussions/2
- [x] https://huggingface.co/hf-internal-testing/tiny-random-mt5/discussions/1
- [x] https://huggingface.co/hf-internal-testing/tiny-random-mt5/discussions/2

### CODE REVIEW

**Comment by ydshieh** [comment]
Thanks, @lewtun 

Let's run the scheduled CI (for ONNX tests) before merge :-)

**Comment by lewtun** [comment]
> Thanks, @lewtun
> 
> Let's run the scheduled CI (for ONNX tests) before merge :-)

Yes, this is still WIP because I discovered some slow tests fail with the new tiny models. Will debug and fix on the model side where necessary :)

File: tests/onnx/test_onnx_v2.py [hunk 1]
```diff
@@ -181,73 +181,70 @@ def test_values_override(self):
...
-    ("mbart", "sshleifer/tiny-mbart"),
-    ("t5", "t5-small"),
-    ("marian", "Helsinki-NLP/opus-mt-en-de"),
-    ("mt5", "google/mt5-base"),
-    ("m2m-100", "facebook/m2m100_418M"),
-    ("blenderbot-small", "facebook/blenderbot_small-90M"),
-    ("blenderbot", "facebook/blenderbot-400M-distill"),
-    ("bigbird-pegasus", "google/bigbird-pegasus-large-arxiv"),
-    ("longt5", "google/long-t5-local-base"),
-    # Disable for now as it causes fatal error `Floating point exception (core dumped)` and the subsequential tests are
```
Comment (lewtun) [inline tests/onnx/test_onnx_v2.py:239]: This problem should now be resolved by use of the tiny model

File: tests/onnx/test_onnx_v2.py [hunk 2]
```diff
@@ -181,73 +181,70 @@ def test_values_override(self):
...
-    ("deit", "facebook/deit-small-patch16-224"),
-    ("beit", "microsoft/beit-base-patch16-224"),
-    ("data2vec-text", "facebook/data2vec-text-base"),
-    ("data2vec-vision", "facebook/data2vec-vision-base"),
-    ("perceiver", "deepmind/language-perceiver", ("masked-lm", "sequence-classification")),
-    ("perceiver", "deepmind/vision-perceiver-conv", ("image-classification",)),
-    ("longformer", "allenai/longformer-base-4096"),
-    ("yolos", "hustvl/yolos-tiny"),
-    ("segformer", "nvidia/segformer-b0-finetuned-ade-512-512"),
+    ("beit", "hf-internal-testing/tiny-random-beit"),
```
Comment (lewtun) [inline tests/onnx/test_onnx_v2.py:184]: I decided to order these alphabetically because it was annoying to find the model names quickly :)

**Comment by ydshieh** [comment]
I left a few comments on Hub PRs. In general, we also like to have small vocab size. But if the current issues persist, no problem for  me to use the tokenizers config or files from the original model checkpoint.

**Comment by lewtun** [comment]
> I left a few comments on Hub PRs. In general, we also like to have small vocab size. But if the current issues persist, no problem for me to use the tokenizers config or files from the original model checkpoint.

Thanks! 

As discussed offline, using a small vocab size `v` in the model config requires that `len(tokenizer) == v`. Otherwise the model cannot run inference because the tokenizer will generate out of vocab input IDs and throw an `index out of range` error.

AFAIK the only way to handle this is to train a tokenizer from scratch on a tiny "corpus" and use the resulting vocab size in the model config. This is simple for fast tokenizers, but somewhat painful for slow ones that don't have the `train_from_iterator()` method. 

In the end it may not be entirely necessary to optimise the model size this way if the resulting "tiny" models are fast enough for our test suite

**Comment by lewtun** [comment]
As discussed internally, we'll revert the changes to the model repos and create dedicated `tiny-random-onnx-x` repos for the ONNX tests

**Comment by lewtun** [comment]
Stable bot begone!

**Comment by ydshieh** [comment]
@gante told me we can use `wip` tag to avoid this bot 😃

**Comment by lewtun** [comment]
Closing in favour of https://github.com/huggingface/transformers/pull/20333

**Pull request closed without merging**